arangodb: 3.4.8 -> 3.10.0

jsoo1 opened the pull request:
###### Description of changes

`arangodb` was out of date (last commited 2018). This brings it to a more current version.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [X] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [X] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [X] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [X] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [X] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

jsoo1 committed:
arangodb/default.nix: format with nixpkgs-fmt

jsoo1 committed:
arangodb_3_9: init at 3.9.3

jsoo1 committed:
arangodb: 3.4.8 -> 3.9.3

jsoo1 committed:
arangodb_3_x: remove unsupported versions

3.{3,4,5} all reached EOL

jsoo1 committed:
arangodb: use gcc 10 (supported version)

jsoo1 committed:
arangodb: remove enableParallelBuilding

It is unused.

jsoo1 committed:
arangodb: drop boost patch

It fails to apply

jsoo1 committed:
arangodb: 3.9.3 -> 3.10.0

jsoo1 commented:
Turns out `3.9.3 -> 3.10.0` wasn't too bad.

jsoo1 committed:
arangodb: note removal of old versions in rl-2211 notes

jsoo1 committed:
arangodb: disable "maintainer mode"

risicle commented on nixos/doc/manual/release-notes/rl-2211.section.md:
```diff
@@ -192,6 +192,8 @@ Available as [services.patroni](options.html#opt-services.patroni.enable).
 
 - The `aws` package has been removed due to being abandoned by the upstream. It is recommended to use `awscli` or `awscli2` instead.
 
+- `arangodb` versions 3.3, 3.4, and 3.5 have been removed because they are at EOL upstream. The default is now 3.10.0.
```
We usually add a `throw` statement to `aliases.nix` for removed packages too.

jsoo1 committed:
aliases.nix: add `throws` for arangodb_3_{3,4,5}

Which were EOL long ago.

risicle commented on pkgs/servers/nosql/arangodb/default.nix:
```diff
@@ -1,68 +1,60 @@
+, which
+}:
+
+gcc10Stdenv.mkDerivation rec {
```
There's an explicit warning in the cmake:

```
CMake Warning at CMakeLists.txt:616 (message):
  ArangoDB requires g++ 11.0 or newer, building with older compiler versions
  is unsupported
```

jsoo1 replied:
I see that but it does seem to compile fine. I'm not sure if they mean that gcc11 is required or that anything else is untested.

risicle replied:
But why fix it to gcc10 in the first place? You haven't explained this.

jsoo1 replied:
You're right! Thank you! Added some commentary.

jsoo1 replied:
Do you think gcc10 is acceptable? I'm not sure what to make of the message (particularly the bit about being "unsupported")

yorickvP replied:
gcc10 seems to work, but the point is, why override it from gcc11, which is the default in nixpkgs?

jsoo1 replied:
Oh! It's just my error! I thought I had checked the stdenv.cc version and apparently I didn't read closely enough.  I'll change this today :)

yorickvP commented:
- #128338
- #128337
- #128336

risicle commented:
Very unlikely, I think @yorickvP 's point is this will fix those issues.

yorickvP commented:
Yep, this upgrade will fix those issues.

jsoo1 committed:
arangodb: annotate gcc10Stdenv reasoning

Co-authored-by: Robert Scott <code@humanleg.org.uk>

jsoo1 commented:
> Yep, this upgrade will fix those issues.

I see, thanks for the clarification!

vcunat commented:
That reasoning doesn't make sense to me.  You'd get gcc11 (which is supported by upstream).  Instead you force gcc10, with an argument speaking against such a change.

jsoo1 committed:
arangodb: stdenvGcc10->stdenv since gcc 11+ is suggested

vcunat commented:
It (now) doesn't build for me on x86_64 NixOS.  Maybe there was some (other) reason to downgrade compiler version.  The errors look like changes around standard strictness, etc.  (complex C++ traces; probably no sense to post them here)

jsoo1 commented:
> It (now) doesn't build for me on x86_64 NixOS. Maybe there was some (other) reason to downgrade compiler version. The errors look like changes around standard strictness, etc. (complex C++ traces; probably no sense to post them here)

Same. I will investigate.

jsoo1 committed:
arangodb: remove unused cmake flags

USE_OPTIMIZE_FOR_ARCHITECTURE was removed without commit in a prior
commit. This is because it was removed upstream:

https://github.com/arangodb/arangodb/commit/cc860c333c7f0c1aed9a2179e3e8b926d92b6df4

This is unfortunate - now /proc/cpuinfo is read unconditionally for
feature detection.

Some other means of feature detection will have to be found.

jsoo1 commented:
I found a new problem - maybe or maybe not related to compile failures - `/proc/cpuinfo` is read unconditionally for architecture feature detection. You turn this off by specifying `-DTARGET_ARCHITECTURE=xyz` up front. However, `-DTARGET_ARCHITECTURE=generic` seems like the best choice and that causes the vendored v8 not to compile.

jsoo1 committed:
arangodb: stdenv->gcc10Stdenv to compile

It seems that gcc11 - though reportedly supported - introduced some
constraint that arangodb fails to compile with.

jsoo1 commented:
I've asked upstream about the gcc 11 compile errors: https://github.com/arangodb/arangodb/issues/17454

jsoo1 committed:
arangodb: additional cmake flags for releases

From https://www.arangodb.com/docs/stable/installation-compiling-debian.html

jsoo1 committed:
arangodb: avoid reading /proc/cpuinfo during build

Avoid non-deterministic builds of ArangoDB proper. Note that this
provides 0 guarantees about the vendored libraries doing anything
similar.

jsoo1 committed:
arangodb: capitalize cmake flag

jsoo1 committed:
arangodb: annotate reasoning for gcc 11->10 downgrade

jsoo1 committed:
arangodb: parameterize target architecture

If -DTARGET_ARCHITECTURE is supplied arango's cmake configuration will
not use /proc/cpuinfo for feature detection. However, `generic` and
`none` options both fail when building the vendored v8.

This workaround provides some defaults, but warns if no override is
provided.

jsoo1 committed:
arangodb: no warning when defaulting target arch

jsoo1 committed:
arangodb: parameterize whether to enable asm optimizations

As far as I can tell, this can cause compile failures when building
the vendored abseil_cpp if choosing a target arch that does not
support sse. This might be possible to determine programmatically, but
it is more flexible to let the user decide.

jsoo1 committed:
arangodb: remove aarch64 support

For which no reliable default target architecture is provided upstream.

jsoo1 committed:
arangodb: document new parameters and aarch64-linux drop

jsoo1 commented:
I decided to keep only `x86_64-linux` as the supported system for `arangodb` since I can't find a reasonable default value for `-DTARGET_ARCHITECTURE` on that platform.  I documented the change and how to alter the new parameters.

jsoo1 commented:
This is probably about as much effort as I can give to this package at the moment. Hopefully what I have is sufficient but I would be happy to see some other eyes on this.

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-ready-for-review/3032/1378

SuperSandro2000 commented on pkgs/servers/nosql/arangodb/default.nix:
```diff
@@ -1,68 +1,81 @@
+  # prevent failing with "cmake-3.13.4/nix-support/setup-hook: line 10: ./3rdParty/rocksdb/RocksDBConfig.cmake.in: No such file or directory"
+  dontFixCmake = true;
+  NIX_CFLAGS_COMPILE = "-Wno-error";
+  preConfigure = "patchShebangs utils";
```
This should really be a multiline string.

jsoo1 replied:
Thank you! https://github.com/NixOS/nixpkgs/pull/200210
